pkg/runtime/wfengine/activity.go

```diff
@@ -78,19 +78,30 @@ func (a *activityActor) InvokeMethod(ctx context.Context, actorID string, method
 	}
 
 	// Try to load activity state. If we find any, that means the activity invocation is a duplicate.
-	if state, err := a.loadActivityState(ctx, actorID, ar.Generation); err != nil {
+	var state activityState
+	if _, err := a.loadActivityState(ctx, actorID, ar.Generation); err != nil {
 		return nil, err
 	} else if state.Generation > 0 {
 		return nil, ErrDuplicateInvocation
 	}
 
+	if methodName == "PurgeWorkflowState" {
+		if err := a.purgeActivityState(ctx, actorID, state); err != nil {
+			return nil, err
+		}
+	}
+
 	// Save the request details to the state store in case we need it after recovering from a failure.
-	state := activityState{
+	state = activityState{
 		Generation:   ar.Generation,
 		EventPayload: ar.HistoryEvent,
 	}
-	if err := a.saveActivityState(ctx, actorID, state); err != nil {
-		return nil, err
+
+	// This conditional block prevents the statestore from re-saving the purged item
+	if methodName != "PurgeWorkflowState" {
```

Comment:
This can (and should) be removed if you add a `return` further up like Alessandro suggested.